### ISSUE TITLE
A Generator's `add_requirements` get lost when used via cpp_stubs or Python's `call()` method

### ISSUE THREAD

**Issue by steven-johnson** [halide/Halide#7044]
When you call `Generator::add_requirement` (in either C++ or Python), we add the requirements to a list in the `Pipeline` we create; then, when lowering, convert those into a series of `AssertStmts` at the top of the block.

A subtle issue here comes in when you want to re-use a Generator at language-composition time, either via `cpp_stubs` or via the equivalent `call()` method in our Python Generators: instead of returning the `Pipeline`, these return the output(s) as a list of `Func`, extracted from the Pipeline... so the requirements are lost, meaning a possibly-vital error check is now missing.

Thinking out loud: perhaps we should:
- move the `add_requirements` method to `Func`
- have `Pipeline` aggregate all the requirements in its output Funcs and emit them at the top of its block (taking care to remove duplicates)

@abadams -- what do you think?

**Comment by abadams** [comment]
That might be weird in the case where the consumer only uses one of the Func outputs, and the requirement refers to things only used by the other Func output.

I guess we could have Func::add_requirement, and Pipeline::add_requirement would just add its requirement to the first output Func? Then they'd get lost if you only use the second output. All output funcs? Then you'd get potentially unwanted requirements again. Hrm.

**Comment by steven-johnson** [comment]
> Hrm

Yeah, the solution isn't immediately obvious, but the status quo definitely feels like a problem too.
